### PR TITLE
Provision for dynamic arguments in the query for data-sharing push via events

### PR DESCRIPTION
**Pull request by knarayan**

### CODE REVIEW

**Comment by VRamakrishna** [comment]
@knarayan Can you add a description of the PR in the first comment above? It's currently empty.

File: core/drivers/fabric-driver/server/fabric-code.ts [hunk 1]
```diff
@@ -77,6 +77,8 @@ const getNetworkGateway = async (networkName: string): Promise<Gateway> => {
 async function invoke(
     query: query_pb.Query,
     networkName: string,
+    funName: string,
```
Comment (VRamakrishna) [inline core/drivers/fabric-driver/server/fabric-code.ts:80]: I'd prefer we call this `funcName` to match our naming conventions elsewhere in this repository.

File: core/drivers/fabric-driver/server/utils.ts [hunk 2]
```diff
@@ -8,6 +8,7 @@ import fs from 'fs';
 import { credentials } from '@grpc/grpc-js';
 import datatransfer_grpc_pb from '@hyperledger-labs/weaver-protos-js/relay/datatransfer_grpc_pb';
 import events_grpc_pb from '@hyperledger-labs/weaver-protos-js/relay/events_grpc_pb';
+import eventsPb from '@hyperledger-labs/weaver-protos-js/common/events_pb';
```
Comment (VRamakrishna) [inline core/drivers/fabric-driver/server/utils.ts:11]: Is this redundant?

File: core/network/fabric-interop-cc/contracts/interop/handle_external_request.go [hunk 3]
```diff
@@ -29,47 +30,75 @@ import (
...
 // 4. Calls application chaincode
 func (s *SmartContract) HandleExternalRequest(ctx contractapi.TransactionContextInterface, b64QueryBytes string) (string, error) {
-	// Ensure that this function cannot be called by a client without relay permissions
-	relayAccessCheck, err := wutils.IsClientRelay(ctx.GetStub())
+	queryBytes, err := base64.StdEncoding.DecodeString(b64QueryBytes)
 	if err != nil {
-		return "", err
+		return "", logThenErrorf("Unable to base64 decode data: %s", err.Error())
 	}
-	if !relayAccessCheck {
```
Comment (VRamakrishna) [inline core/network/fabric-interop-cc/contracts/interop/handle_external_request.go:37]: Why did you delete this?

File: core/network/fabric-interop-cc/contracts/interop/handle_external_request.go [hunk 4]
```diff
@@ -29,47 +30,75 @@ import (
 // 3. Checks the access control policy for the requester and view address is met
 // 4. Calls application chaincode
 func (s *SmartContract) HandleExternalRequest(ctx contractapi.TransactionContextInterface, b64QueryBytes string) (string, error) {
-	// Ensure that this function cannot be called by a client without relay permissions
-	relayAccessCheck, err := wutils.IsClientRelay(ctx.GetStub())
```
Comment (VRamakrishna) [inline core/network/fabric-interop-cc/contracts/interop/handle_external_request.go:33]: Why did you delete this?

File: core/network/fabric-interop-cc/contracts/interop/handle_external_request.go [hunk 5]
```diff
@@ -29,47 +30,75 @@ import (
...
-		return "", errors.New(errorMessage)
+		return "", logThenErrorf("Unable to unmarshal query: %s", err.Error())
 	}
+
+	queryAddress := query.Address
+	dynamicArgCount := strings.Count(query.Address, ":?")
+	if dynamicArgCount > 1 {
+		return "", logThenErrorf("Expected 1 dynamic argument in the event query address, but found %d", dynamicArgCount)
+	} else if dynamicArgCount == 1 {
+		queryAddress = strings.Replace(query.Address, ":?", ":"+dynamicQueryArg, 1)
```
Comment (VRamakrishna) [inline core/network/fabric-interop-cc/contracts/interop/handle_external_request.go:62]: Add a TODO comment above this saying that this is stopgap logic and will be reviewed and possibly changed later.

File: core/network/fabric-interop-cc/contracts/interop/handle_external_request.go [hunk 6]
```diff
@@ -29,47 +30,75 @@ import (
...
+		fmt.Println("There is 1 dynamic argument in the event query address, queryArg: ", dynamicQueryArg)
+	} else {
+		fmt.Println("There are no dynamic arguments in the event query address, queryArg: ", dynamicQueryArg)
+	}
+
+	resp, err := handleRequest(s, ctx, query, queryAddress)
+	return resp, err
+}
+
+// handleRequest function requests that come from external networks.
```
Comment (VRamakrishna) [inline core/network/fabric-interop-cc/contracts/interop/handle_external_request.go:72]: Say: "This function handles requests that originate in external networks and have come through relays".

File: samples/fabric/fabric-cli/src/data/event_sub_sample.json [hunk 7]
```diff
@@ -8,16 +8,35 @@
...
         },
         "event_publication_spec": {
-            "ccArgs": [],
+            "ccArgs": ["keyType", ""],
             "chaincodeId": "simplestate",
             "driverId": "network1",
             "ccFunc": "Create",
             "channelId": "mychannel",
-            "replaceArgIndex": "0"
+            "replaceArgIndex": 1
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/src/data/event_sub_sample.json:16]: How and where is this property used?

File: samples/fabric/fabric-cli/src/data/event_sub_sample.json [hunk 8]
```diff
@@ -8,16 +8,35 @@
             "transactionFunc": "create"
         },
         "event_publication_spec": {
-            "ccArgs": [],
+            "ccArgs": ["keyType", ""],
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/src/data/event_sub_sample.json:11]: What does "keyType" denote?

File: samples/fabric/fabric-cli/src/data/event_sub_sample.json [hunk 9]
```diff
@@ -27,16 +46,35 @@
             "transactionFunc": "create"
         },
         "event_publication_spec": {
-            "ccArgs": [],
+            "ccArgs": ["keyClass", ""],
```
Comment (VRamakrishna) [inline samples/fabric/fabric-cli/src/data/event_sub_sample.json:49]: What does "keyClass" denote?

File: core/drivers/fabric-driver/server/fabric-code.ts [hunk 10]
```diff
@@ -77,6 +77,8 @@ const getNetworkGateway = async (networkName: string): Promise<Gateway> => {
...
+    dynamicArg: Buffer | undefined
```
Comment (sanvenDev) [inline core/drivers/fabric-driver/server/fabric-code.ts:81]: I think to make it optional, we can use `?` here.
Otherwise we still will need to pass `undefined` here. If we use `?`, then by default it will be undefined.

File: samples/fabric/fabric-cli/src/data/event_sub_sample.json (hunk 7 above)
Comment (sanvenDev) [reply samples/fabric/fabric-cli/src/data/event_sub_sample.json:16]: This is used when destination driver submits a transaction. 
I remember this was modified in previous PR of Krishna, I don't know why the main didn't had it yet.

File: core/drivers/fabric-driver/server/fabric-code.ts (hunk 1 above)
Comment (knarayan) [reply core/drivers/fabric-driver/server/fabric-code.ts]: Done

File: core/drivers/fabric-driver/server/utils.ts (hunk 2 above)
Comment (knarayan) [reply core/drivers/fabric-driver/server/utils.ts]: Removed now Rama

File: core/network/fabric-interop-cc/contracts/interop/handle_external_request.go (hunk 6 above)
Comment (knarayan) [reply core/network/fabric-interop-cc/contracts/interop/handle_external_request.go]: Done

File: samples/fabric/fabric-cli/src/data/event_sub_sample.json (hunk 7 above)
Comment (knarayan) [reply samples/fabric/fabric-cli/src/data/event_sub_sample.json:16]: The index here need to be an integer.

File: samples/fabric/fabric-cli/src/data/event_sub_sample.json (hunk 8 above)
Comment (knarayan) [reply samples/fabric/fabric-cli/src/data/event_sub_sample.json:11]: It's just a name for the key imported from remote network.

File: samples/fabric/fabric-cli/src/data/event_sub_sample.json (hunk 9 above)
Comment (knarayan) [reply samples/fabric/fabric-cli/src/data/event_sub_sample.json:49]: It's just a name for the key imported from remote network.

File: core/drivers/fabric-driver/server/fabric-code.ts (hunk 10 above)
Comment (knarayan) [reply core/drivers/fabric-driver/server/fabric-code.ts]: Done Sandeep.

**Review by sanvenDev — APPROVED** [review]

**Review by sanvenDev — CHANGES_REQUESTED** [review]
Data sharing tests are failing.

**Review by sanvenDev — APPROVED** [review]
